Introduce an Incus connection plugin

stgraber opened the pull request:
##### SUMMARY
This introduces a new connection plugin for Incus.

This is extremely heavily based on the great work of @mattclay with the LXD plugin.

On top of a general rename of anything LXD to Incus, this also cleans up a bunch of the wording to be inclusive of virtual machines. It also now uses the `--quiet` flag to avoid weird artifacts during file transfers.

Closes #7376 

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
plugins/connection/incus

felixfontein commented on changelogs/fragments/7726-incus-plugin.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - "incus - add a new connection plugin for Incus"
```
Please remove this file. New modules and plugins are announced by their `version_added` value and `short_description`, they don't need an explicit changelog fragment.

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+    short_description: Run tasks in Incus instances via the Incus CLI
+    description:
+        - Run commands or put/fetch files to an existing Incus instance using Incus CLI
+    version_added: "2.0"
```
The next minor release of community.general is 8.2.0:
```suggestion
    version_added: "8.2.0"
```

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+    connection: incus
+    short_description: Run tasks in Incus instances via the Incus CLI
+    description:
+        - Run commands or put/fetch files to an existing Incus instance using Incus CLI
```
Please note that all `description`s should contain paragraphs, which consist out of proper sentences.
```suggestion
        - Run commands or put/fetch files to an existing Incus instance using Incus CLI.
```

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        self._host = self._play_context.remote_addr.split(".")[0]
```
`remote_addr` should be accessed with `self.get_option('remote_addr')`. Also that value might change over time, so it should be done when needed, and not in advance.

bcoca replied:
this was here in the old plugin as it used outdated play_context to get initial value, no need for it anymore

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        self._host = self._play_context.remote_addr.split(".")[0]
+        self._incus_cmd = find_executable("incus")
```
Please use ansible-core's code for this (`from ansible.module_utils.common.process import get_bin_path`).

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+        if not self._incus_cmd:
+            raise AnsibleError("incus command not found in PATH")
+
+        if self._play_context.remote_user is not None and self._play_context.remote_user != 'root':
```
In general `self._play_context` should be avoided if possible. Since `remote_user` is not an option for this plugin, I would simply remove this check.

felixfontein reviewed (commented):
Thanks for your contribution! Please note that new plugins must come with tests (unit and/or integration tests).

Also please note that the lxd connection plugin is relatively old, so several things it does might not be acceptable for new code anymore.

stgraber commented:
Thanks!

I marked the PR as draft now while I address those comments and the remaining CI failures.

I'll have to see what can be done for a useful test on a connection plugin, hopefully that doesn't require more python than I can handle (I'm a Go dev these days).

bcoca commented:
@stgraber i'll help as much as i can, I was considering writing this plugin since the license change to lxd .. but you beat me to it!

bcoca commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+            - name: ansible_incus_executable
+      remote:
+        description:
+            - name of the Incus remote to use
```
I would explain what a 'remote' is, consider that many users might not know about incus or the way the term is used in that context. I would say the same for 'project'.

bcoca commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,155 @@
+        - Run commands or put/fetch files to an existing Incus instance using Incus CLI
+    version_added: "2.0"
+    options:
+      remote_addr:
```
This can be confusing as `project + instance_name` is unique but the instance_name itself is not not unique across projects but for Ansible `inventory_hostname` is the unique identifier.

Does it make sense to have  `instance_name@project` as the `inventory_hostname` and then split on that? 
or are most instances going to be `inventory_hostname`(instance name) at `default` (project)?

bcoca replied:
this is probably something to be addressed with the inventory plugin also.

stgraber replied:
I suspect most will be inventory_hostname at default project. It's only when you're dealing with larger scale deployments that the projects come at play.

bcoca replied:
k, leave as is for now, we can address on inventory when `project !=default` and/or `remote != local`. We can come up with conventions there for `inventory_hostname` and just populate the `ansible_incus_` variables.

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,168 @@
+            - name: ansible_incus_executable
+      remote:
+        description:
+            - The name of the Incus remote to use (per "incus remote list").
```
```suggestion
            - The name of the Incus remote to use (per C(incus remote list)).
```
(https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#format-macros-within-module-documentation)

stgraber commented:
Alright, that looks like the CI stuff and I believe review comments addressed?

@bcoca @felixfontein anything I missed in there before I go looking at how tests work?

stgraber commented:
So I just copy/pasted the integration test for LXD into a new Incus one.

It doesn't really seem to be doing anything useful though and I don't know exactly how much we need it to in order to satisfy the requirement.

felixfontein commented:
> It doesn't really seem to be doing anything useful though and I don't know exactly how much we need it to in order to satisfy the requirement.

It does run a few standardized tests for connection plugins, both with and without pipelining. What happened if you run that test (since it doesn't run in CI)? Does it pass?

stgraber commented:
@felixfontein okay, I got the integration test to behave. Had to do a few tweaks to the one LXD has in place as I don't see how it could have worked. You also obviously need to run this on a machine that has Incus installed, so can't be tested inside the usual Docker test container.

stgraber commented:
```
root@v1:~/ansible_collections/community/general# incus admin init --auto
root@v1:~/ansible_collections/community/general# incus launch images:ubuntu/22.04 ubuntu-2204
Creating ubuntu-2204
Starting ubuntu-2204                        
root@v1:~/ansible_collections/community/general# ansible-test integration connection_incus --allow-unsupported
WARNING: A Python 3 interpreter was found at "/usr/bin/python3", yet a matching pip was not found.
Running connection_incus integration test script
++ python -c 'from os import path; print(path.basename(path.abspath(path.dirname('\''./runme.sh'\''))).replace('\''connection_'\'', '\'''\''))'
+ group=incus
+ cd ../connection
+ INVENTORY=../connection_incus/test_connection.inventory
+ ./test.sh -e target_hosts=incus -e action_prefix= -e local_tmp=/tmp/ansible-local -e remote_tmp=/tmp/ansible-remote
+ '[' -f ../connection_incus/test_connection.inventory ']'
+ ansible-playbook test_connection.yml -i ../connection_incus/test_connection.inventory -e target_hosts=incus -e action_prefix= -e local_tmp=/tmp/ansible-local -e remote_tmp=/tmp/ansible-remote
[WARNING]: Collection community.general does not support Ansible version 2.10.8
[WARNING]: running playbook inside collection community.general
[WARNING]: The `junit_xml` python module is not installed. Disabling the `junit` callback plugin.

PLAY [incus] **************************************************************************************************************************************************

TASK [raw with unicode arg and output] ************************************************************************************************************************
changed: [incus-pipelining]

TASK [check output of raw with unicode arg and output] ********************************************************************************************************
ok: [incus-pipelining] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [create local file with unicode filename and content] ****************************************************************************************************
changed: [incus-pipelining]

TASK [remove remote file with unicode filename and content] ***************************************************************************************************
ok: [incus-pipelining]

TASK [create remote directory with unicode name] **************************************************************************************************************
changed: [incus-pipelining]

TASK [copy local file with unicode filename and content] ******************************************************************************************************
changed: [incus-pipelining]

TASK [remove local file with unicode filename and content] ****************************************************************************************************
changed: [incus-pipelining]

TASK [fetch remote file with unicode filename and content] ****************************************************************************************************
changed: [incus-pipelining]

TASK [remove local temp file] *********************************************************************************************************************************
changed: [incus-pipelining]

TASK [remove remote temp file] ********************************************************************************************************************************
changed: [incus-pipelining]

TASK [ansible.builtin.wait_for_connection] ********************************************************************************************************************
[WARNING]: Reset is not implemented for this connection
ok: [incus-pipelining]

PLAY [incus] **************************************************************************************************************************************************

TASK [raw with unicode arg and output] ************************************************************************************************************************
changed: [incus-no-pipelining]

TASK [check output of raw with unicode arg and output] ********************************************************************************************************
ok: [incus-no-pipelining] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [create local file with unicode filename and content] ****************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove remote file with unicode filename and content] ***************************************************************************************************
ok: [incus-no-pipelining]

TASK [create remote directory with unicode name] **************************************************************************************************************
changed: [incus-no-pipelining]

TASK [copy local file with unicode filename and content] ******************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove local file with unicode filename and content] ****************************************************************************************************
changed: [incus-no-pipelining]

TASK [fetch remote file with unicode filename and content] ****************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove local temp file] *********************************************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove remote temp file] ********************************************************************************************************************************
changed: [incus-no-pipelining]

TASK [ansible.builtin.wait_for_connection] ********************************************************************************************************************
[WARNING]: Reset is not implemented for this connection
ok: [incus-no-pipelining]

PLAY RECAP ****************************************************************************************************************************************************
incus-no-pipelining        : ok=11   changed=8    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
incus-pipelining           : ok=11   changed=8    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

+ grep -E ' 2\.(9|10|11|12|13)\.'
+ grep ansible
+ ansible --version
ansible 2.10.8
+ LC_ALL=C
+ LANG=C
+ ansible-playbook test_connection.yml -i ../connection_incus/test_connection.inventory -e target_hosts=incus -e action_prefix= -e local_tmp=/tmp/ansible-local -e remote_tmp=/tmp/ansible-remote
[WARNING]: Collection community.general does not support Ansible version 2.10.8
[WARNING]: running playbook inside collection community.general
[WARNING]: The `junit_xml` python module is not installed. Disabling the `junit` callback plugin.

PLAY [incus] **************************************************************************************************************************************************

TASK [raw with unicode arg and output] ************************************************************************************************************************
changed: [incus-pipelining]

TASK [check output of raw with unicode arg and output] ********************************************************************************************************
ok: [incus-pipelining] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [create local file with unicode filename and content] ****************************************************************************************************
changed: [incus-pipelining]

TASK [remove remote file with unicode filename and content] ***************************************************************************************************
ok: [incus-pipelining]

TASK [create remote directory with unicode name] **************************************************************************************************************
changed: [incus-pipelining]

TASK [copy local file with unicode filename and content] ******************************************************************************************************
changed: [incus-pipelining]

TASK [remove local file with unicode filename and content] ****************************************************************************************************
changed: [incus-pipelining]

TASK [fetch remote file with unicode filename and content] ****************************************************************************************************
changed: [incus-pipelining]

TASK [remove local temp file] *********************************************************************************************************************************
changed: [incus-pipelining]

TASK [remove remote temp file] ********************************************************************************************************************************
changed: [incus-pipelining]

TASK [ansible.builtin.wait_for_connection] ********************************************************************************************************************
[WARNING]: Reset is not implemented for this connection
ok: [incus-pipelining]

PLAY [incus] **************************************************************************************************************************************************

TASK [raw with unicode arg and output] ************************************************************************************************************************
changed: [incus-no-pipelining]

TASK [check output of raw with unicode arg and output] ********************************************************************************************************
ok: [incus-no-pipelining] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [create local file with unicode filename and content] ****************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove remote file with unicode filename and content] ***************************************************************************************************
ok: [incus-no-pipelining]

TASK [create remote directory with unicode name] **************************************************************************************************************
changed: [incus-no-pipelining]

TASK [copy local file with unicode filename and content] ******************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove local file with unicode filename and content] ****************************************************************************************************
changed: [incus-no-pipelining]

TASK [fetch remote file with unicode filename and content] ****************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove local temp file] *********************************************************************************************************************************
changed: [incus-no-pipelining]

TASK [remove remote temp file] ********************************************************************************************************************************
changed: [incus-no-pipelining]

TASK [ansible.builtin.wait_for_connection] ********************************************************************************************************************
[WARNING]: Reset is not implemented for this connection
ok: [incus-no-pipelining]

PLAY RECAP ****************************************************************************************************************************************************
incus-no-pipelining        : ok=11   changed=8    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
incus-pipelining           : ok=11   changed=8    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

WARNING: Reviewing previous 1 warning(s):
WARNING: A Python 3 interpreter was found at "/usr/bin/python3", yet a matching pip was not found.
root@v1:~/ansible_collections/community/general# echo $?
0
```

bcoca commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,168 @@
+        """ close the connection (nothing to do here) """
+        super(Connection, self).close()
+
+        self._connected = False
```
can we add a transport_test method?  just to check basic connectivity works, something like `incus shell true` or equivalent?

stgraber replied:
Do you have any example for that? I'm not seeing any hit for `transport_test` in the community.general repo.

stgraber replied:
FWIW, I just checked and the current integration tests exercise `file pull`, `file push` and `exec`, which are the three commands exposed by the connection plugin.

```
root@v1:~# grep /usr/bin/incus.*exec trace | wc -l
1738
root@v1:~# grep /usr/bin/incus.*file.*pull trace | wc -l
62
root@v1:~# grep /usr/bin/incus.*file.*push trace | wc -l
252
root@v1:~# 
```

This is based on a `forkstat` trace to get all the calls to the incus CLI tool.

bcoca replied:
it was an addition we did to `wait_for_connection` and has not really caught on yet

stgraber replied:
Do we still need that given that the integration tests exercise all 3 code paths we care about?

felixfontein commented:
> @felixfontein okay, I got the integration test to behave. Had to do a few tweaks to the one LXD has in place as I don't see how it could have worked.

Since this test didn't run in CI it's possible that it has been broken for a long time without anyone noticing.

> You also obviously need to run this on a machine that has Incus installed, so can't be tested inside the usual Docker test container.

We also have a few VMs in CI, so it would be possible to run tests. The question is how hard (both from amount of work and amount of time it needs in every CI run) it is to setup Incus.

felixfontein commented on plugins/connection/incus.py:
```diff
@@ -0,0 +1,168 @@
+    options:
+      remote_addr:
+        description:
+            - Instance identifier
```
```suggestion
            - The instance identifier.
```

stgraber replied:
Fixed!

stgraber commented:
> We also have a few VMs in CI, so it would be possible to run tests. The question is how hard (both from amount of work and amount of time it needs in every CI run) it is to setup Incus.

Probably best to wait for major distros to have native packages available.

It'd be possible to do currently on Ubuntu LTS, Debian, Gentoo and on Fedora but only Gentoo would have distro-native packages, the rest all relies on installing 3rd party repositories at this stage.

Given we know the test actually works now and does exercise all aspects of the code, it feels like we're in a much better place than we ever were with the LXD plugin :)

stgraber commented:
Anything else needed before this one can be merged?

felixfontein reviewed (approved):
Looks good to me. If nobody objects, I'll merge this in ~a week.

felixfontein commented:
@stgraber thanks for your contribution!
@bcoca thanks for reviewing!
